### ISSUE TITLE
GPU memory is always raising

### ISSUE THREAD

**Issue by haduoken** [hkchengrex/Cutie#86]
I use 3080TI with 12G GPU mem, when use the interactive demo with an video file, it can process to the end (about 5k frame)
but when I change to use the script demo mode, the GPU mem is always raising, and it process about 200 frame, then be killed

**Comment by hkchengrex** [comment]
I cannot reproduce this problem. Did you modify any code?

**Comment by haduoken** [comment]
` # add to ref frame
        with torch.no_grad() and torch.cuda.amp.autocast():
            from cutie.inference.inference_core import InferenceCore
            from cutie.utils.get_default_model import get_default_model

            cutie = get_default_model()
            processor = InferenceCore(cutie, cfg=cutie.cfg)
            # the processor matches the shorter edge of the input to this size
            # you might want to experiment with different sizes, -1 keeps the original size
            processor.max_internal_size = 120

            image = torch.from_numpy(img)
            image = image.permute(2, 0, 1) / 255
            image = image.to(device)

            masks = masks.long()
            masks = masks.to(device)

            output_prob = processor.step(image, masks, objects=[1])
            mask = processor.output_prob_to_mask(output_prob)
            mask = mask == 1

            frame = 1
            while True:
                ok, img = cap.read()
                if not ok:
                    print('video done')
                    break
                img = cv2.resize(img, (640, 360))

                image = torch.from_numpy(img)
                image = image.permute(2, 0, 1) / 255
                image = image.to(device)

                output_prob = processor.step(image)

                mask = processor.output_prob_to_mask(output_prob)
                mask = mask == 1

                # render mask
                image = draw_segmentation_masks(image.cpu(), mask.cpu().unsqueeze(0))
                image = image.permute(1, 2, 0).numpy()
                cv2.imshow('cam', image)
                cv2.waitKey(1)

                print(f'process frame {frame}')
                frame += 1`

**Comment by hkchengrex** [comment]
This portion of code looks alright to me. Are you sure the memory increase is not caused by other things happening in your program?

**Comment by haduoken** [comment]
thanks again for reply

I save the mask file and process the video with script_demo.py alone, It works pretty good

I will close this issue later (up to 10h),  when I find the bug, I will share it here 

```
import os

import torch
from torchvision.transforms.functional import to_tensor
from PIL import Image
import numpy as np

from cutie.inference.inference_core import InferenceCore
from cutie.utils.get_default_model import get_default_model

import cv2


@torch.inference_mode()
@torch.cuda.amp.autocast()
def main():
    cap = cv2.VideoCapture('/home/kilox/output.mp4')
    # obtain the Cutie model with default parameters -- skipping hydra configuration
    cutie = get_default_model()
    # Typically, use one InferenceCore per video
    processor = InferenceCore(cutie, cfg=cutie.cfg)
    # the processor matches the shorter edge of the input to this size
    # you might want to experiment with different sizes, -1 keeps the original size
    processor.max_internal_size = 480

    mask = Image.open('/home/kilox/yijie/Cutie/examples/masks/bike/00000.png')
    palette = mask.getpalette()

    mask = np.load('mask.npy')

    mask = mask.astype(np.int32)

    # the number of objects is determined by counting the unique values in the mask
    # common mistake: if the mask is resized w/ interpolation, there might be new unique values
    objects = np.unique(np.array(mask))
    # background "0" does not count as an object
    objects = objects[objects != 0].tolist()

    mask = torch.from_numpy(np.array(mask)).cuda()

    for ti in range(5000):
        # load the image as RGB; normalization is done within the model
        ok, img = cap.read()

        image = torch.from_numpy(img)
        image = image.permute(2, 0, 1) / 255
        image = image.cuda()

        # image = Image.open(os.path.join(image_path, image_name))
        # image = to_tensor(image).cuda().float()

        if ti == 0:
            # if mask is passed in, it is memorized
            # if not all objects are specified, we propagate the unspecified objects using memory
            output_prob = processor.step(image, mask, objects=objects)
        else:
            # otherwise, we propagate the mask from memory
            output_prob = processor.step(image)

        # convert output probabilities to an object mask
        mask = processor.output_prob_to_mask(output_prob)

        # visualize prediction
        mask = Image.fromarray(mask.cpu().numpy().astype(np.uint8))
        mask.putpalette(palette)
        # mask.show()  # or use mask.save(...) to save it somewhere
        print(f'process frame {ti}')


main()

```

**Comment by haduoken** [comment]
ok, I find it

the following code only acquire context of second expr
```
with torch.no_grad() and torch.cuda.amp.autocast():
```

so we should change to 
```
  with torch.no_grad():
      with torch.cuda.amp.autocast():
             xxx
```


the following code only print 
```
bar enter
xx
bar exit
```
```
class foo:
    def __init__(self) -> None:
        self.prev = False

    def __enter__(self) -> None:
        print('foo enter')

    def __exit__(self, exc_type, exc_value, traceback) -> None:
        print('foo exit')


class bar:
    def __init__(self) -> None:
        self.prev = False

    def __enter__(self) -> None:
        print('bar enter')

    def __exit__(self, exc_type, exc_value, traceback) -> None:
        print('bar exit')


with foo() and bar():
    print('xx')

```

**Comment by hkchengrex** [comment]
Thank you for the update.
I think it might be `with foo(), bar():`? https://stackoverflow.com/questions/3024925/create-a-with-block-on-several-context-managers
